include/pinctrl.h

```diff
@@ -0,0 +1,811 @@
+ *
+ * Pin control interface for
+ * - pin control: get pin control properties
+ * - pin configuration: configure a pins electronic properties
+ * - pin multiplexing: reuse the same pin for different purposes
+ * @{
+ */
+
+#include <pinctrl_common.h>
+
+#include <zephyr/types.h>
+#include <device.h>
+
+
+/**
+ * @def PINCTRL_CONFIG
+ * @brief Create a pinctrl config value from pinctrl configuration defines.
+ *
+ * The pinctrl configuration defines shall be the
+ * @ref pinctrl_interface_config_defs "pinctrl configuration defines"
+ * as given in the pinctrl_common.h file.
+ */
+#define PINCTRL_CONFIG(...) _PINCTRL_CONFIG0(__VA_ARGS__)
+/**
+ * @def _PINCTRL_CONFIG0
+ * @private
+ */
+#define _PINCTRL_CONFIG0(...)                                   \
+	_PINCTRL_CONFIG32(__VA_ARGS__,                          \
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, \
+		0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
+/**
+ * @def _PINCTRL_CONFIG32
+ * @private
+ */
+#define _PINCTRL_CONFIG32(c0,  c1,  c2,  c3,  c4,  c5,  c6,  c7,  \
+			  c8,  c9,  c10, c11, c12, c13, c14, c15, \
+			  c16, c17, c18, c19, c20, c21, c22, c23, \
+			  c24, c25, c26, c27, c28, c29, c30, c31, \
+			  ...)                                    \
+	(c0  | c1  | c2  | c3  | c4  | c5  | c6  | c7  |          \
+	 c8  | c9  | c10 | c11 | c12 | c13 | c14 | c15 |          \
+	 c16 | c17 | c18 | c19 | c20 | c21 | c22 | c23 |          \
+	 c24 | c25 | c26 | c17 | c28 | c29 | c30 | c31)
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct pinctrl_control_api {
+	u16_t (*get_pins_count)(struct device *dev);
+	int (*get_groups_count)(struct device *dev);
+	int (*get_group_pins)(struct device *dev, u16_t group,
+			      u16_t *pins, u16_t *num_pins);
+	u16_t (*get_states_count)(struct device *dev);
+	int (*get_state_group)(struct device *dev, u16_t state, u16_t *group);
+	u16_t (*get_functions_count)(struct device *dev);
+	int (*get_function_group)(struct device *dev, u16_t func,
+				  const char *name, u16_t *group);
+	int (*get_function_groups)(struct device *dev, u16_t func,
+				   u16_t *groups, u16_t *num_groups);
+	int (*get_function_state)(struct device *dev, u16_t func,
+				  const char *name, u16_t *state);
+	int (*get_function_states)(struct device *dev, u16_t func,
+				   u16_t *states, u16_t *num_states);
+	int (*get_device_function)(struct device *dev, struct device *client,
+				   u16_t *func);
+	int (*get_gpio_range)(struct device *dev, struct device *gpio,
+			      u32_t gpio_pin, u16_t *pin, u16_t *base_pin,
+			      u8_t *num_pins);
+};
+
+struct pinctrl_config_api {
+	int (*get)(struct device *dev, u16_t pin, u32_t *config);
+	int (*set)(struct device *dev, u16_t pin, u32_t config);
+	int (*group_get)(struct device *dev, u16_t group,
+			 u32_t *configs, u16_t *num_configs);
+	int (*group_set)(struct device *dev, u16_t group,
+			 const u32_t *configs, u16_t num_configs);
+};
+
+struct pinctrl_mux_api {
+#if CONFIG_PINCTRL_MUX_REQUEST
```

Comment:
maybe I missed it, but it there a need for this? I did not see any usage (maybe I missed it)

It would too easy to lock a pin.

Also, if this has to be kept, owner a const char: not good. Just make it const void *, so the caller put whatever it wants (its own dev pointer, or else).

Reply:
The usage is what you mentioned. Detect when a pin is locked by some other user. It is more for runtime use if you reconfigure the pin functions. But a double request can also happen due to a wrong DTS as already happened to me. If everything works you can switch it off - that is why the CONFIG_PINCTRL_MUX_REQUEST exists.

The driver implements a string compare here. How to compare a void pointer?

Reply:
The name is being used here as a kind of a handle right? So anything would do. A const void * would. Just compare the pointer itself (not what's inside obviously).

Anyway, you configure a pin in on go or? (set/group_set function below) If so, would a lock through a semaphore be sufficient in the set/group_set functions then?